tracer/src/Datadog.Trace/Debugger/Instrumentation/Collections/ProbeDataCollection.cs

```diff
@@ -42,52 +42,84 @@ internal static ProbeMetadataCollection Instance
+                // So we essentially perform the ProbeId checking twice to make sure the `ProbeData` returned is indeed related
+                // to the probeId given. Of course, one tick later it could be invalidated too, but at least we willreturn `ProbeData` that
+                // is related to the given `probeId` and not of another probe.
+                return ref GetProbeDataAtIndex(index, probeId);
             }
 
             // Create a new one at the given index
             lock (ItemsLocker)
             {
-                if (IsIndexExists(index))
+                if (IsIndexExists(index, probeId))
                 {
-                    return true;
+                    // Read the comment above, of the same block of code.
+                    return ref GetProbeDataAtIndex(index, probeId);
                 }
 
                 EnlargeCapacity(index);
 
                 if (Log.IsEnabled(Vendors.Serilog.Events.LogEventLevel.Debug))
                 {
-                    Log.Debug($"{nameof(ProbeMetadataCollection)}.{nameof(TryCreateProbeMetadataIfNotExists)}: Creating a new probe metadata info for index = {index}, Items.Length = {Items.Length}");
+                    Log.Debug($"{nameof(ProbeDataCollection)}.{nameof(TryCreateProbeDataIfNotExists)}: Creating a new probe metadata info for index = {index}, Items.Length = {Items.Length}");
                 }
 
                 var processor = ProbeExpressionsProcessor.Instance.Get(probeId);
 
                 if (processor == null)
                 {
-                    return false;
+                    return ref ProbeData.Empty;
                 }
 
                 var sampler = ProbeRateLimiter.Instance.GerOrAddSampler(probeId);
 
                 Items[index] = new ProbeData(probeId, sampler, processor);
             }
 
-            return true;
+            return ref GetProbeDataAtIndex(index, probeId);
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         protected override bool IsEmpty(ref ProbeData payload)
         {
             return payload == default;
         }
+
+        /// <summary>
+        /// Assumes `index` exists.
+        /// <seealso cref="IsIndexExists"/>
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private ref ProbeData GetProbeDataAtIndex(int index, string probeId)
+        {
+            ref var probeData = ref Items[index];
+            if (probeData.ProbeId == probeId)
+            {
+                return ref probeData;
+            }
+
+            return ref ProbeData.Empty;
+        }
+
+        /// <summary>
+        /// In the native side we are trying to reuse indices. When a probe gets removed, it's associated probe data indices
+        /// will be reused in subsequent probes instrumentation. To make sure the index embedded in the instrumentation is not reused,
+        /// we compare also the ProbeId in the location.
```

Comment:
```suggestion
        /// In the native side we attempt to reuse indices; when a probe gets removed, its associated probe data index
        /// will be reused when a new probe later gets add. To make sure the index embedded in the instrumentation is not stale,
        /// we compare the ProbeId at that index with the one we were expecting.
```

Reply:
Done in 90e0ba24cfd43eab1fca79e8a887db33e9ddf49e.